src/deque.cr

```diff
@@ -182,21 +182,83 @@ class Deque(T)
   # a             # => Deque{"a", "c"}
   # ```
   def delete(obj)
-    delete_if { |i| i == obj }
+    match = internal_delete { |i| i == obj }
+    !match.nil?
   end
 
-  def delete_if
-    found = false
+  # Modifies `self`, keeping only the elements in the collection for which the
+  # passed block returns `true`. Returns `self`.
+  #
+  # ```
+  # a = Deque{1, 6, 2, 4, 8}
+  # a.select! { |x| x > 3 }
+  # a # => Deque{6, 4, 8}
+  # ```
+  #
+  # See also: `Deque#select`.
+  def select!
+    reject! { |elem| !yield(elem) }
+  end
+
+  # Modifies `self`, keeping only the elements in the collection for which
+  # `pattern === element`.
+  #
+  # ```
+  # ary = [1, 6, 2, 4, 8]
+  # ary.select!(3..7)
+  # ary # => [6, 4]
+  # ```
+  #
+  # See also: `Deque#select`.
+  def select!(pattern)
+    self.select! { |elem| pattern === elem }
+  end
+
+  # Modifies `self`, deleting the elements in the collection for which the
+  # passed block returns `true`. Returns `self`.
+  #
+  # ```
+  # a = Deque{1, 6, 2, 4, 8}
+  # a.reject! { |x| x > 3 }
+  # a # => Deque{1, 2}
+  # ```
+  #
+  # See also: `Deque#reject`.
+  def reject!
+    internal_delete { |e| yield e }
+    self
```

Comment:
Let's make return self or nil as in Hash.

Ref: https://github.com/crystal-lang/crystal/pull/9856#issuecomment-723141092

Reply:
I'd prefer to change `Hash#reject!` to be non-nilable instead. There's really not much reason to return `nil` and it breaks method chaining. If you need to know whether an item was removed, you can keep track of whether the block ever returns `false` as demonstrated in https://github.com/crystal-lang/crystal/issues/9879#issuecomment-721898344
That clearly communicates what's going on and avoids having to do some nasty nil-checks when chaning calls. I doubt there are many use cases for the nilable return type anyway.

Reply:
Changing `Hash#reject!` to return always self would be a silent breaking-change.

Ruby returns self or nil on `{Array, Hash}#reject!`.

`tap` can be used to perform chaining if wanted. `c.tap(&.reject!(...)).foo`

Reply:
> Changing Hash#reject! to return always self would be a silent breaking-change.

Yes, that's unfortunate. We don't necessarily have to change it, at least not right now. But I would not introduce the same weird return type for new methods.

> tap can be used to perform chaining if wanted. c.tap(&.reject!(...)).foo

This clearly shows that it doesn't make any sense to have `#reject!` return `self` as an indicator for an item being deleted. Returning `Bool` would be sufficient if such a return value is desired. Otherwise it should just be `self` always.

Reply:
Should we then update `Array#reject!` to also return `self | nil`?
I don't see a particularly good reason for the inconsistency between these interfaces besides _least astonishment_ coming from ruby. I think the return type of `self` is better purely for the benefit of simply chaining methods.

Reply:
Ok, let's keep `reject! : self` then. We will change the `hash#reject!` in another PR.... I hope this breaking-change will not haunt us later...